File: Scripts/modelling/00_report_output_dictionary.py
from __future__ import annotations

import csv
import textwrap
from pathlib import Path
from typing import Any


SECTION_TITLE = "CSV Output Dictionary"
WRAP_WIDTH = 100


def col(name: str, meaning: str, formula: str | None = None) -> dict[str, str]:
    entry = {"name": name, "meaning": meaning}
    if formula:
        entry["formula"] = formula
    return entry


def dynamic_tail(label: str, meaning: str, formula: str | None = None) -> dict[str, str]:
    entry = {"label": label, "meaning": meaning}
    if formula:
        entry["formula"] = formula
    return entry


def csv_output_spec(
    path: str | Path,
    row_grain: str,
    columns: list[dict[str, str]],
    dynamic_columns: dict[str, str] | None = None,
) -> dict[str, Any]:
    return {
        "path": Path(path),
        "row_grain": row_grain,
        "columns": columns,
        "dynamic_columns": dynamic_columns,
    }


def read_csv_header(path: str | Path) -> list[str]:
    with open(path, newline="") as handle:
        return next(csv.reader(handle))


def _wrap_line(prefix: str, text: str) -> list[str]:
    wrapped = textwrap.wrap(
        text,
        width=WRAP_WIDTH,
        initial_indent=prefix,
        subsequent_indent=" " * len(prefix),
        break_long_words=False,
        break_on_hyphens=False,
    )
    return wrapped or [prefix.rstrip()]


def _validate_and_get_header(spec: dict[str, Any]) -> list[str]:
    path = Path(spec["path"])
    actual_header = read_csv_header(path)
    fixed_header = [entry["name"] for entry in spec["columns"]]

    if spec.get("dynamic_columns") is None:
        if actual_header != fixed_header:
            raise ValueError(
                f"Header mismatch for {path}. Expected {fixed_header}, found {actual_header}"
            )
    else:
        prefix = actual_header[: len(fixed_header)]
        if prefix != fixed_header:
            raise ValueError(
                f"Header prefix mismatch for {path}. Expected {fixed_header}, found {actual_header}"
            )
        if len(actual_header) <= len(fixed_header):
            raise ValueError(
                f"Expected dynamic columns after fixed prefix {fixed_header} in {path}, found none."
            )

    return actual_header


def render_csv_output_dictionary(specs: list[dict[str, Any]]) -> str:
    lines = [SECTION_TITLE, ""]

    for index, spec in enumerate(specs):
        path = Path(spec["path"])
        header = _validate_and_get_header(spec)

        lines.append(path.name)
        lines.append(f"Path: {path}")
        lines.append(f"Row grain: {spec['row_grain']}")
        lines.append("Columns:")

        for entry in spec["columns"]:
            text = f"{entry['name']}: {entry['meaning']}"
            if entry.get("formula"):
                text += f" Formula: {entry['formula']}"
            lines.extend(_wrap_line("- ", text))

        dynamic_spec = spec.get("dynamic_columns")
        if dynamic_spec is not None:
            dynamic_names = header[len(spec["columns"]) :]
            text = (
                f"{dynamic_spec['label']} ({len(dynamic_names)} total): {dynamic_spec['meaning']} "
                f"Exact column names: {', '.join(dynamic_names)}"
            )
            if dynamic_spec.get("formula"):
                text += f" Formula: {dynamic_spec['formula']}"
            lines.extend(_wrap_line("- ", text))

        if index != len(specs) - 1:
            lines.append("")

    return "\n".join(lines)


def strip_existing_csv_output_dictionary(report_text: str) -> str:
    stripped = report_text.rstrip()
    if stripped == SECTION_TITLE:
        return ""

    marker = f"\n{SECTION_TITLE}\n"
    position = stripped.find(marker)
    if position != -1:
        return stripped[:position].rstrip()

    if stripped.startswith(f"{SECTION_TITLE}\n"):
        return ""

    return stripped


def append_csv_output_dictionary_to_text(report_text: str, specs: list[dict[str, Any]]) -> str:
    base_text = strip_existing_csv_output_dictionary(report_text)
    dictionary_text = render_csv_output_dictionary(specs)
    if base_text:
        return f"{base_text}\n\n{dictionary_text}\n"
    return f"{dictionary_text}\n"


def build_report_text(report_lines: list[str], specs: list[dict[str, Any]]) -> str:
    return append_csv_output_dictionary_to_text("\n".join(report_lines), specs)

File: Scripts/plotting_r/00_report_output_dictionary.R
read_csv_header <- function(path) {
  first_line <- readLines(path, n = 1L, warn = FALSE)
  strsplit(first_line, ",", fixed = TRUE)[[1]]
}

wrap_prefixed_line <- function(prefix, text, width = 100) {
  wrapped <- strwrap(
    text,
    width = width,
    exdent = nchar(prefix),
    simplify = FALSE
  )[[1]]
  if (length(wrapped) == 0) {
    return(prefix)
  }
  c(
    paste0(prefix, wrapped[1]),
    if (length(wrapped) > 1) {
      paste0(strrep(" ", nchar(prefix)), wrapped[-1])
    } else {
      character(0)
    }
  )
}

validate_report_output_dictionary <- function(specs) {
  for (spec in specs) {
    header <- read_csv_header(spec$path)
    fixed_header <- vapply(spec$columns, `[[`, character(1), "name")

    if (is.null(spec$dynamic_columns)) {
      if (!identical(header, fixed_header)) {
        stop(
          sprintf(
            "Header mismatch for %s. Expected %s, found %s",
            spec$path,
            paste(fixed_header, collapse = ", "),
            paste(header, collapse = ", ")
          )
        )
      }
    } else {
      prefix <- header[seq_along(fixed_header)]
      if (!identical(prefix, fixed_header)) {
        stop(
          sprintf(
            "Header prefix mismatch for %s. Expected %s, found %s",
            spec$path,
            paste(fixed_header, collapse = ", "),
            paste(header, collapse = ", ")
          )
        )
      }
      if (length(header) <= length(fixed_header)) {
        stop(sprintf("Expected dynamic columns after fixed prefix in %s", spec$path))
      }
    }
  }
}

render_csv_output_dictionary <- function(specs) {
  validate_report_output_dictionary(specs)
  lines <- c("CSV Output Dictionary", "")

  for (index in seq_along(specs)) {
    spec <- specs[[index]]
    header <- read_csv_header(spec$path)

    lines <- c(
      lines,
      basename(spec$path),
      paste("Path:", spec$path),
      paste("Row grain:", spec$row_grain),
      "Columns:"
    )

    for (column in spec$columns) {
      text <- paste0(column$name, ": ", column$meaning)
      if (!is.null(column$formula) && nzchar(column$formula)) {
        text <- paste0(text, " Formula: ", column$formula)
      }
      lines <- c(lines, wrap_prefixed_line("- ", text))
    }

    if (!is.null(spec$dynamic_columns)) {
      dynamic_header <- header[(length(spec$columns) + 1):length(header)]
      dynamic_text <- paste0(
        spec$dynamic_columns$label,
        " (",
        length(dynamic_header),
        " total): ",
        spec$dynamic_columns$meaning,
        " Exact column names: ",
        paste(dynamic_header, collapse = ", ")
      )
      if (!is.null(spec$dynamic_columns$formula) && nzchar(spec$dynamic_columns$formula)) {
        dynamic_text <- paste0(dynamic_text, " Formula: ", spec$dynamic_columns$formula)
      }
      lines <- c(lines, wrap_prefixed_line("- ", dynamic_text))
    }

    if (index != length(specs)) {
      lines <- c(lines, "")
    }
  }

  lines
}

strip_existing_csv_output_dictionary <- function(report_lines) {
  marker_index <- match("CSV Output Dictionary", report_lines)
  if (is.na(marker_index)) {
    return(trimws(report_lines, which = "right"))
  }
  trimws(report_lines[seq_len(marker_index - 1L)], which = "right")
}

append_csv_output_dictionary_to_lines <- function(report_lines, specs) {
  base_lines <- strip_existing_csv_output_dictionary(report_lines)
  dictionary_lines <- render_csv_output_dictionary(specs)
  c(base_lines, "", dictionary_lines)
}
